### PR TITLE
PLUGIN-78 - Zendesk Batch Source Single Object plugin

### PR DESCRIPTION
**Pull request by rostyslavmorozov**
WIKI: https://wiki.cask.co/display/CE/Zendesk+plugins
JIRA: https://issues.cask.co/browse/PLUGIN-78

Included:
- Single Object Batch Source
- Multi Object Batch Source

@albertshau, @CuriousVini, @mikkin could you please review this PR?
Thanks for your attention.

### CODE REVIEW

File: src/main/java/io/cdap/plugin/zendesk/source/common/config/BaseZendeskSourceConfig.java [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+  public void validate(FailureCollector collector) {
+    IdUtils.validateReferenceName(referenceName, collector);
+    if (!containsMacro(PROPERTY_ADMIN_EMAIL)
+      && !EmailValidator.getInstance().isValid(adminEmail)) {
+      collector.addFailure(String.format("'%s' is not a valid email.", adminEmail), null)
+        .withConfigProperty(PROPERTY_ADMIN_EMAIL);
+    }
+    if (!Strings.isNullOrEmpty(objectsToSkip)
+      && getObjects().isEmpty()) {
+      collector.addFailure("All objects are skipped.", null)
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/zendesk/source/common/config/BaseZendeskSourceConfig.java]: what is the corrective action in this case?

File: src/main/java/io/cdap/plugin/zendesk/source/batch/ZendeskBatchMultiSource.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+  public ZendeskBatchMultiSource(ZendeskBatchSourceConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    FailureCollector failureCollector = pipelineConfigurer.getStageConfigurer().getFailureCollector();
+    config.validate(failureCollector);
+    failureCollector.getOrThrowException();
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/zendesk/source/batch/ZendeskBatchMultiSource.java:59]: no need to call this method in configurePipeline

File: src/main/java/io/cdap/plugin/zendesk/source/batch/ZendeskRecordReader.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+import org.apache.hadoop.mapreduce.RecordReader;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Iterator;
+
+/**
+ * RecordReader implementation, which reads object from Zendesk using
+ * zendesk-java-client.
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/zendesk/source/batch/ZendeskRecordReader.java]: where are we using zendesk-java-client?

File: src/main/java/io/cdap/plugin/zendesk/source/batch/ZendeskBatchSource.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+import io.cdap.cdap.etl.api.batch.BatchSourceContext;
+import io.cdap.plugin.common.LineageRecorder;
+import io.cdap.plugin.zendesk.source.common.config.BaseZendeskSourceConfig;
+import org.apache.hadoop.io.NullWritable;
+
+import java.util.Collections;
+import java.util.stream.Collectors;
+
+/**
+ * Plugin returns records from Zendesk API using provided by user object.
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/zendesk/source/batch/ZendeskBatchSource.java]: Source plugin to read data from Zendesk.

File: src/main/java/io/cdap/plugin/zendesk/source/batch/ZendeskSplit.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+  }
+
+  @Override
+  public void readFields(DataInput in) throws IOException {
+    subdomain = in.readUTF();
+    object = in.readUTF();
+  }
+
+  @Override
+  public long getLength() throws IOException, InterruptedException {
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/zendesk/source/batch/ZendeskSplit.java]: nit: can remove the exceptions

File: src/main/java/io/cdap/plugin/zendesk/source/batch/ZendeskSplit.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    return 0;
+  }
+
+  @Override
+  public String[] getLocations() throws IOException, InterruptedException {
```
Comment (CuriousVini) [inline src/main/java/io/cdap/plugin/zendesk/source/batch/ZendeskSplit.java]: nit: can remove the exceptions

File: pom.xml [hunk 7]
```diff
@@ -0,0 +1,457 @@
...
+      <groupId>com.google.inject</groupId>
+      <artifactId>guice</artifactId>
+      <version>${guice.version}</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>com.cloudbees.thirdparty</groupId>
+      <artifactId>zendesk-java-client</artifactId>
+      <version>${zendesk-java-client.version}</version>
+      <scope>test</scope>
```
Comment (CuriousVini) [inline pom.xml:252]: why is this test scope? Are we not relying on zendesk client?

File: src/main/java/io/cdap/plugin/zendesk/source/common/config/BaseZendeskSourceConfig.java (hunk 1 above)
Comment (rostyslavmorozov) [reply src/main/java/io/cdap/plugin/zendesk/source/common/config/BaseZendeskSourceConfig.java]: Will add this one:
`Make sure 'Objects to Pull' and 'Objects to Skip' fields don't hold same values.`
Is it ok?

File: src/main/java/io/cdap/plugin/zendesk/source/batch/ZendeskBatchMultiSource.java (hunk 2 above)
Comment (rostyslavmorozov) [reply src/main/java/io/cdap/plugin/zendesk/source/batch/ZendeskBatchMultiSource.java:59]: In this case not all validation errors will be shown after clicking 'Validate' button. Should I remove it?

File: src/main/java/io/cdap/plugin/zendesk/source/batch/ZendeskRecordReader.java (hunk 3 above)
Comment (rostyslavmorozov) [reply src/main/java/io/cdap/plugin/zendesk/source/batch/ZendeskRecordReader.java]: Only in tests. Will remove it.

File: pom.xml (hunk 7 above)
Comment (rostyslavmorozov) [reply pom.xml:252]: Zendesk Java client covers not all endpoints (approximately 40-50%) and it is not extensible.

File: src/main/java/io/cdap/plugin/zendesk/source/batch/ZendeskBatchSource.java (hunk 4 above)
Comment (rostyslavmorozov) [reply src/main/java/io/cdap/plugin/zendesk/source/batch/ZendeskBatchSource.java]: Done.

File: src/main/java/io/cdap/plugin/zendesk/source/batch/ZendeskSplit.java (hunk 5 above)
Comment (rostyslavmorozov) [reply src/main/java/io/cdap/plugin/zendesk/source/batch/ZendeskSplit.java]: Done.

File: src/main/java/io/cdap/plugin/zendesk/source/batch/ZendeskSplit.java (hunk 6 above)
Comment (rostyslavmorozov) [reply src/main/java/io/cdap/plugin/zendesk/source/batch/ZendeskSplit.java]: Done.

**Comment by rostyslavmorozov** [comment]
@CuriousVini I made code changes and commented some items, could you please review?

**Pull request merged**